### ISSUE TITLE
Secure Authentication for Online IDE

### ISSUE THREAD

**Issue by iyannsch** [ls1intum/Artemis#8763]
Twofold situation: 
- We need to protect users from having their token/credentials stolen
- We need to protect the system from users that abuse / hide between a Theia session to cheat

Situation JWT: 
- User´s JWT Token is saved as HTTPOnly and cannot be read from JS
- Artemis sets the Cookie and reads it on API invocation
- From the Theia instance, we need to access Artemis APIs and websockets

Situation Git Token:
- Theia needs to clone and push the user`s repo
- We don`t want the user to insert their username and password all over again
- We could add functionality to Artemis that allows token-based git interaction and provide Theia with such a token on startup
- Potential security problem: This token can modify the user’s submission

Situation Theia Session Sharing:
- Theia Sessions can be protected using OIDC, but are not in our base use-case
- Users can share their Theia-Session URL allowing multiple users to work in one workspace
- Two problems:
    - 1) Any token/credential saved in the Theia context can leak to third-parties authorized to access the session
    - 2) Without authentication in Theia, we don’t know how many users are working together and submit using a reverse-proxy-like setup.
- Potential solution: Authentication users with OIDC before granting them session access generates traces at the KeyCloak OID-Provider allowing fraud detection just like the existing session monitoring 

With OIDC:
- 1 Login in Artemis
- 1 Login in Theia for Instance Access
- 1 Login in Plugin for Artemis Access 

Probleme:
- Selbst wenn auf gleicher URL, können wir es wrsl nicht nutzen, da wir Nutzer-Sessions 

Open: 
- Kann man jetzt nicht auch einfach den „Git Token“ (also das, was IntelliJ speichert) teilen, um unbemerkt von mehreren Geräten auf Repos zu pushen?
- Wenn Theia einen HTTPOnly Token kriegt/hat (z.B. bei auf gleicher Domain wie Artemis), ist der dann für Authentifizierung gegenüber Artemis aus einem Plugin verfügbar?

# Feature Proposal
> Spec Version 0.2.0

### Problem  
> Describe the problem that is tackled in this issue

### Motivation 
> Describe the motivation WHY the problem needs solving. Specify the affected users/roles.

### Existing (Problematic) Solution / System 
> What is the current solution (if there is one)? What is the problem with the current solution? 
> You may include a UML Model here 

### Proposed System 
> What would the ideal solution look like?  

### Requirements 
> Describe the Functional and Non-Functional Requirements of the feature. Stick to the INVEST methodology! 
> 1. FR: <Title>: <Description> 
>
> 1. NFR: <FURPS+ Category>: <Title>: <Description>

### Analysis Object Model 
> What are the involved Analysis Objects? 

### Dynamic Behavior 
> Include dynamic models (Activity Diagram, State Chart Diagram, Communication Diagram) here to outline the dynamic nature of the PROBLEM 

### Subsystem Decomposition
> Show the involved subsystems and their interfaces. Make sure to describe the APIs that you add/change in detail. Model the DTOs you intend to (re)use or change! 

### Persistent Data Management
> Describe the Database changes you intend to make.
> Outline new configuration options you plan to introduce
> Describe all other data persistence mechanisms you may use.

### Access Control / Security Aspects 
> Describe the access control considerations for your feature

### Other Design Decisions
> Potential topics to discuss here include: WebSockets, testing strategies.

## UI/UX Design
> Screenshots of the final UI mockups (mandatory): Please include screenshots to provide a clear and persistent visual reference of the design. 
> Link to the design mockup (optional): Additionally, you may include a link to the live design mockup (e.g., Figma, Sketch) for a more interactive view. Note that this link is supplementary and should not replace the required screenshots.

**Comment by krusche** [comment]
> JWT is not built to generate additional tokens seamlessly.

I do not agree with this statement. Users can login on multiple browsers, each time a new JWT token is generated. So in some sense, you could argue that one user can have arbitrary many JWT tokens at the same time.

The Artemis client / browser stores the token in the cookie so that links to attachments can be shared and opened in separate tabs without the web application being active. 

However, this mechanism is not set in stone. If there is a specific requirement to change it, we can discuss it.

Another alternative would be to host the client part of the Online IDE using the same URL as the Artemis instance based on a micro frontend approach. Then, the online IDE could directly access the JWT token within the cookie.
